File: FX/OAuth2Client/IOS/libOAuth2Client.cs
//  Import of libOAuth2Client ()
//  Frameworks:
//  Targets: armv7, armv7s, arm64
//  Dep fx:rtl, Foundation
//  Dep libs:OAuth2Client
namespace 
{
    public class OAuth2Client.__Global
    {
        public static NSString NXOAuth2ErrorDomain;
        public static NSInteger NXOAuth2InvalidRequestErrorCode;
        public static NSInteger NXOAuth2InvalidClientErrorCode;
        public static NSInteger NXOAuth2UnauthorizedClientErrorCode;
        public static NSInteger NXOAuth2RedirectURIMismatchErrorCode;
        public static NSInteger NXOAuth2AccessDeniedErrorCode;
        public static NSInteger NXOAuth2UnsupportedResponseTypeErrorCode;
        public static NSInteger NXOAuth2InvalidScopeErrorCode;
        public static NSInteger NXOAuth2CouldNotRefreshTokenErrorCode;
        public static NSString NXOAuth2HTTPErrorDomain;
        public static NSString NXOAuth2AccountStoreErrorKey;
        public static NSString NXOAuth2ClientConnectionContextTokenRequest;
        public static NSString NXOAuth2ClientConnectionContextTokenRefresh;
        public static NSString NXOAuth2ConnectionDidStartNotification;
        public static NSString NXOAuth2ConnectionDidEndNotification;
        public static NSString NXOAuth2AccountStoreDidFailToRequestAccessNotification;
        public static NSString NXOAuth2AccountStoreAccountsDidChangeNotification;
        public static NSString NXOAuth2AccountStoreNewAccountUserInfoKey;
        public static NSString kNXOAuth2AccountStoreConfigurationClientID;
        public static NSString kNXOAuth2AccountStoreConfigurationSecret;
        public static NSString kNXOAuth2AccountStoreConfigurationAuthorizeURL;
        public static NSString kNXOAuth2AccountStoreConfigurationTokenURL;
        public static NSString kNXOAuth2AccountStoreConfigurationRedirectURL;
        public static NSString kNXOAuth2AccountStoreConfigurationScope;
        public static NSString kNXOAuth2AccountStoreConfigurationTokenType;
        public static NSString kNXOAuth2AccountStoreConfigurationTokenRequestHTTPMethod;
        public static NSString kNXOAuth2AccountStoreConfigurationAdditionalAuthenticationParameters;
        public static NSString kNXOAuth2AccountStoreConfigurationCustomHeaderFields;
        public static NSString kNXOAuth2AccountStoreAccountType;
        public static NSString NXOAuth2AccountDidChangeUserDataNotification;
        public static NSString NXOAuth2AccountDidChangeAccessTokenNotification;
        public static NSString NXOAuth2AccountDidLoseAccessTokenNotification;
        public static NSString NXOAuth2AccountDidFailToGetAccessTokenNotification;
        public static const Int32 NXOAuth2ConnectionDebug;
    }

    public enum OAuth2Client.NXOAuth2TrustMode: Int32
    {
        NXOAuth2TrustModeAnyCertificate,
        AnyCertificate,
        NXOAuth2TrustModeSystem,
        System,
        NXOAuth2TrustModeSpecificCertificate,
        SpecificCertificate
    }

    public interface OAuth2Client.INXOAuth2TrustDelegate
    {
        OAuth2Client.NXOAuth2TrustMode connection(OAuth2Client.NXOAuth2Connection connection) trustModeForHostname(NSString hostname);
        NSArray connection(OAuth2Client.NXOAuth2Connection connection) trustedCertificatesForHostname(NSString hostname);
    }

    public interface OAuth2Client.INXOAuth2ClientDelegate
    {
        void oauthClientNeedsAuthentication(OAuth2Client.NXOAuth2Client client);
        void oauthClientDidGetAccessToken(OAuth2Client.NXOAuth2Client client);
        void oauthClientDidLoseAccessToken(OAuth2Client.NXOAuth2Client client);
        void oauthClientDidRefreshAccessToken(OAuth2Client.NXOAuth2Client client);
        void oauthClient(OAuth2Client.NXOAuth2Client client) didFailToGetAccessTokenWithError(NSError error);
    }

    public interface OAuth2Client.INXOAuth2ConnectionDelegate
    {
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didReceiveResponse(NSURLResponse response);
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didFinishWithData(NSData data);
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didFailWithError(NSError error);
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didReceiveData(NSData data);
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didSendBytes(UInt64 bytesSend) ofTotal(UInt64 bytesTotal);
        void oauthConnection(OAuth2Client.NXOAuth2Connection connection) didReceiveRedirectToURL(NSURL redirectURL);
    }

    public class OAuth2Client.NXOAuth2Client: NSObject
    {
        protected BOOL authenticating;
        protected BOOL persistent;
        protected NSString clientId;
        protected NSString clientSecret;
        protected NSSet desiredScope;
        protected NSString userAgent;
        protected NSString assertion;
        protected NSString keyChainGroup;
        protected NSURL authorizeURL;
        protected NSURL tokenURL;
        protected NSString tokenType;
        protected OAuth2Client.NXOAuth2Connection authConnection;
        protected OAuth2Client.NXOAuth2AccessToken accessToken;
        protected NSMutableArray waitingConnections;
        protected NSInteger refreshConnectionDidRetryCount;
        protected OAuth2Client.INXOAuth2ClientDelegate @delegate;

        public BOOL authenticating { get; }
        public NSString clientId { get; }
        public NSString clientSecret { get; }
        public NSString tokenType { get; }
        public NSDictionary additionalAuthenticationParameters { get; set; }
        public NSDictionary customHeaderFields { get; set; }
        public NSSet desiredScope { get; set; }
        public NSString tokenRequestHTTPMethod { get; set; }
        public NSString userAgent { get; set; }
        public NSString acceptType { get; set; }
        public OAuth2Client.NXOAuth2AccessToken accessToken { get; set; }
        public OAuth2Client.INXOAuth2ClientDelegate @delegate { get; set; }
        public BOOL persistent { get; set; }
        public id initWithClientID(NSString clientId) clientSecret(NSString clientSecret) authorizeURL(NSURL authorizeURL) tokenURL(NSURL tokenURL) @delegate(OAuth2Client.INXOAuth2ClientDelegate @delegate);
        public id initWithClientID(NSString clientId) clientSecret(NSString clientSecret) authorizeURL(NSURL authorizeURL) tokenURL(NSURL tokenURL) accessToken(OAuth2Client.NXOAuth2AccessToken accessToken) keyChainGroup(NSString keyChainGroup) persistent(BOOL shouldPersist) @delegate(OAuth2Client.INXOAuth2ClientDelegate @delegate);
        public id initWithClientID(NSString clientId) clientSecret(NSString clientSecret) authorizeURL(NSURL authorizeURL) tokenURL(NSURL tokenURL) accessToken(OAuth2Client.NXOAuth2AccessToken accessToken) tokenType(NSString tokenType) keyChainGroup(NSString keyChainGroup) persistent(BOOL shouldPersist) @delegate(OAuth2Client.INXOAuth2ClientDelegate @delegate);
        public BOOL openRedirectURL(NSURL URL);
        public NSURL authorizationURLWithRedirectURL(NSURL redirectURL);
        public void authenticateWithClientCredentials();
        public void authenticateWithUsername(NSString username) password(NSString password);
        public void authenticateWithAssertionType(NSURL assertionType) assertion(NSString assertion);
        public void requestAccess();
        public void refreshAccessToken();
        public void refreshAccessTokenAndRetryConnection(OAuth2Client.NXOAuth2Connection retryConnection);
    }

    public class OAuth2Client.NXOAuth2AccessToken: NSObject
    {
        private NSString accessToken;
        private NSString refreshToken;
        private NSString tokenType;
        private NSDate expiresAt;
        private NSSet scope;
        private NSString responseBody;

        public NSString accessToken { get; }
        public NSString refreshToken { get; }
        public NSString tokenType { get; }
        public NSDate expiresAt { get; }
        public BOOL doesExpire { get; }
        public BOOL hasExpired { get; }
        public NSSet scope { get; }
        public NSString responseBody { get; }
        public static id tokenWithResponseBody(NSString responseBody);
        public static id tokenWithResponseBody(NSString responseBody) tokenType(NSString tokenType);
        public id initWithAccessToken(NSString accessToken);
        public id initWithAccessToken(NSString accessToken) refreshToken(NSString refreshToken) expiresAt(NSDate expiryDate);
        public id initWithAccessToken(NSString accessToken) refreshToken(NSString refreshToken) expiresAt(NSDate expiryDate) scope(NSSet scope);
        public id initWithAccessToken(NSString accessToken) refreshToken(NSString refreshToken) expiresAt(NSDate expiryDate) scope(NSSet scope) responseBody(NSString responseBody);
        public id initWithAccessToken(NSString accessToken) refreshToken(NSString refreshToken) expiresAt(NSDate expiryDate) scope(NSSet scope) responseBody(NSString responseBody) tokenType(NSString tokenType);
        public void restoreWithOldToken(OAuth2Client.NXOAuth2AccessToken oldToken);
        public static id tokenFromDefaultKeychainWithServiceProviderName(NSString provider);
        public void storeInDefaultKeychainWithServiceProviderName(NSString provider);
        public void removeFromDefaultKeychainWithServiceProviderName(NSString provider);
    }

    public     public void (NSURLResponse response, NSData responseData, NSError error);

    public     public void (UInt64 bytesSend, UInt64 bytesTotal);

    public class OAuth2Client.NXOAuth2Connection: NSObject
    {
        private NSURLConnection connection;
        private NSMutableURLRequest request;
        private NSURLResponse response;
        private NSDictionary requestParameters;
        private NSMutableData data;
        private BOOL savesData;
        private id context;
        private NSDictionary userInfo;
        private OAuth2Client.NXOAuth2Client client;
        private OAuth2Client.INXOAuth2ConnectionDelegate @delegate;
        private OAuth2Client.NXOAuth2ConnectionResponseHandler responseHandler;
        private OAuth2Client.NXOAuth2ConnectionSendingProgressHandler sendingProgressHandler;
        private BOOL sendConnectionDidEndNotification;

        public OAuth2Client.INXOAuth2ConnectionDelegate @delegate { get; set; }
        public NSData data { get; }
        public BOOL savesData { get; set; }
        public Int64 expectedContentLength { get; }
        public NSURLResponse response { get; }
        public NSInteger statusCode { get; }
        public id context { get; set; }
        public NSDictionary userInfo { get; set; }
        public OAuth2Client.NXOAuth2Client client { get; }
        public id initWithRequest(NSMutableURLRequest request) requestParameters(NSDictionary requestParameters) oauthClient(OAuth2Client.NXOAuth2Client client) sendingProgressHandler(OAuth2Client.NXOAuth2ConnectionSendingProgressHandler sendingProgressHandler) responseHandler(OAuth2Client.NXOAuth2ConnectionResponseHandler responseHandler);
        public id initWithRequest(NSMutableURLRequest request) requestParameters(NSDictionary requestParameters) oauthClient(OAuth2Client.NXOAuth2Client client) @delegate(OAuth2Client.INXOAuth2ConnectionDelegate @delegate);
        public void cancel();
        public void retry();
    }

    public class OAuth2Client.NXOAuth2FileStreamWrapper: NSObject
    {
        protected NSInputStream stream;
        protected UInt64 contentLength;
        protected NSString fileName;
        protected NSString contentType;

        public NSInputStream stream { get; }
        public UInt64 contentLength { get; }
        public NSString fileName { get; }
        public NSString contentType { get; set; }
        public static id wrapperWithStream(NSInputStream stream) contentLength(UInt64 contentLength);
        public id initWithStream(NSInputStream stream) contentLength(UInt64 contentLength);
        public static id wrapperWithStream(NSInputStream stream) contentLength(UInt64 contentLength) fileName(NSString fileName);
        public id initWithStream(NSInputStream stream) contentLength(UInt64 contentLength) fileName(NSString fileName);
    }

    public class OAuth2Client.NXOAuth2PostBodyStream: NSInputStream
    {
        protected NSString boundary;
        protected NSArray contentStreams;
        protected NSInputStream currentStream;
        protected NSUInteger streamIndex;
        protected UInt64 numBytesTotal;

        public id initWithParameters(NSDictionary postParameters);

        public NSString boundary { get; }
        public UInt64 length { get; }
    }

    public     public OAuth2Client.NXOAuth2TrustMode (OAuth2Client.NXOAuth2Connection connection, NSString hostname);

    public     public NSArray (NSString hostname);

    public     public void (NSURL preparedURL);

    public class OAuth2Client.NXOAuth2AccountStore: NSObject
    {
        private NSMutableDictionary pendingOAuthClients;
        private NSMutableDictionary accountsDict;
        private NSMutableDictionary configurations;
        private NSMutableDictionary trustModeHandler;
        private NSMutableDictionary trustedCertificatesHandler;

        public static id sharedStore();

        public NSArray accounts { get; }
        public NSArray accountsWithAccountType(NSString accountType);
        public OAuth2Client.NXOAuth2Account accountWithIdentifier(NSString identifier);
        public void setClientID(NSString aClientID) secret(NSString aSecret) authorizationURL(NSURL anAuthorizationURL) tokenURL(NSURL aTokenURL) redirectURL(NSURL aRedirectURL) forAccountType(NSString anAccountType);
        public void setClientID(NSString aClientID) secret(NSString aSecret) scope(NSSet theScope) authorizationURL(NSURL anAuthorizationURL) tokenURL(NSURL aTokenURL) redirectURL(NSURL aRedirectURL) keyChainGroup(NSString aKeyChainGroup) forAccountType(NSString anAccountType);
        public void setClientID(NSString aClientID) secret(NSString aSecret) scope(NSSet theScope) authorizationURL(NSURL anAuthorizationURL) tokenURL(NSURL aTokenURL) redirectURL(NSURL aRedirectURL) keyChainGroup(NSString aKeyChainGroup) tokenType(NSString aTokenType) forAccountType(NSString anAccountType);
        public void setConfiguration(NSDictionary configuration) forAccountType(NSString accountType);
        public NSDictionary configurationForAccountType(NSString accountType);
        public void setTrustModeHandlerForAccountType(NSString accountType) block(OAuth2Client.NXOAuth2TrustModeHandler handler);
        public OAuth2Client.NXOAuth2TrustModeHandler trustModeHandlerForAccountType(NSString accountType);
        public void setTrustedCertificatesHandlerForAccountType(NSString accountType) block(OAuth2Client.NXOAuth2TrustedCertificatesHandler handler);
        public OAuth2Client.NXOAuth2TrustedCertificatesHandler trustedCertificatesHandlerForAccountType(NSString accountType);
        public void requestAccessToAccountWithType(NSString accountType);
        public void requestAccessToAccountWithType(NSString accountType) withPreparedAuthorizationURLHandler(OAuth2Client.NXOAuth2PreparedAuthorizationURLHandler aPreparedAuthorizationURLHandler);
        public void requestAccessToAccountWithType(NSString accountType) username(NSString username) password(NSString password);
        public void requestAccessToAccountWithType(NSString accountType) assertionType(NSURL assertionType) assertion(NSString assertion);
        public void requestClientCredentialsAccessWithType(NSString accountType);
        public void addAccount(OAuth2Client.NXOAuth2Account account);
        public void removeAccount(OAuth2Client.NXOAuth2Account account);
        public BOOL handleRedirectURL(NSURL URL);
    }

    public class OAuth2Client.NXOAuth2Account: NSObject
    {
        private NSString accountType;
        private NSString identifier;
        private void userData;
        private OAuth2Client.NXOAuth2Client oauthClient;
        private OAuth2Client.NXOAuth2AccessToken accessToken;

        public NSString accountType { get; }
        public NSString identifier { get; }
        public void userData { get; set; }
        public OAuth2Client.NXOAuth2Client oauthClient { get; }
        public OAuth2Client.NXOAuth2AccessToken accessToken { get; }
    }

    public class OAuth2Client.NXOAuth2Request: NSObject
    {
        private NSDictionary parameters;
        private NSURL resource;
        private NSString requestMethod;
        private OAuth2Client.NXOAuth2Account account;
        private OAuth2Client.NXOAuth2Connection connection;
        private OAuth2Client.NXOAuth2Request me;

        public static void performMethod(NSString method) onResource(NSURL resource) usingParameters(NSDictionary parameters) withAccount(OAuth2Client.NXOAuth2Account account) sendProgressHandler(OAuth2Client.NXOAuth2ConnectionSendingProgressHandler progressHandler) responseHandler(OAuth2Client.NXOAuth2ConnectionResponseHandler responseHandler);
        public id initWithResource(NSURL url) method(NSString method) parameters(NSDictionary parameter);

        public OAuth2Client.NXOAuth2Account account { get; set; }
        public NSString requestMethod { get; set; }
        public NSURL resource { get; set; }
        public NSDictionary parameters { get; set; }
        public NSURLRequest signedURLRequest();
        public void performRequestWithSendingProgressHandler(OAuth2Client.NXOAuth2ConnectionSendingProgressHandler progressHandler) responseHandler(OAuth2Client.NXOAuth2ConnectionResponseHandler responseHandler);
        public void cancel();
    }
}

File: FX/OAuth2Client/IOS/libOAuth2Client.pas
//  Import of libOAuth2Client ()
//  Frameworks:
//  Targets: armv7, armv7s, arm64
//  Dep fx:rtl, Foundation
//  Dep libs:OAuth2Client
namespace ;

interface

type
  OAuth2Client.__Global = public class
  public
    class var NXOAuth2ErrorDomain: NSString;
    class var NXOAuth2InvalidRequestErrorCode: NSInteger;
    class var NXOAuth2InvalidClientErrorCode: NSInteger;
    class var NXOAuth2UnauthorizedClientErrorCode: NSInteger;
    class var NXOAuth2RedirectURIMismatchErrorCode: NSInteger;
    class var NXOAuth2AccessDeniedErrorCode: NSInteger;
    class var NXOAuth2UnsupportedResponseTypeErrorCode: NSInteger;
    class var NXOAuth2InvalidScopeErrorCode: NSInteger;
    class var NXOAuth2CouldNotRefreshTokenErrorCode: NSInteger;
    class var NXOAuth2HTTPErrorDomain: NSString;
    class var NXOAuth2AccountStoreErrorKey: NSString;
    class var NXOAuth2ClientConnectionContextTokenRequest: NSString;
    class var NXOAuth2ClientConnectionContextTokenRefresh: NSString;
    class var NXOAuth2ConnectionDidStartNotification: NSString;
    class var NXOAuth2ConnectionDidEndNotification: NSString;
    class var NXOAuth2AccountStoreDidFailToRequestAccessNotification: NSString;
    class var NXOAuth2AccountStoreAccountsDidChangeNotification: NSString;
    class var NXOAuth2AccountStoreNewAccountUserInfoKey: NSString;
    class var kNXOAuth2AccountStoreConfigurationClientID: NSString;
    class var kNXOAuth2AccountStoreConfigurationSecret: NSString;
    class var kNXOAuth2AccountStoreConfigurationAuthorizeURL: NSString;
    class var kNXOAuth2AccountStoreConfigurationTokenURL: NSString;
    class var kNXOAuth2AccountStoreConfigurationRedirectURL: NSString;
    class var kNXOAuth2AccountStoreConfigurationScope: NSString;
    class var kNXOAuth2AccountStoreConfigurationTokenType: NSString;
    class var kNXOAuth2AccountStoreConfigurationTokenRequestHTTPMethod: NSString;
    class var kNXOAuth2AccountStoreConfigurationAdditionalAuthenticationParameters: NSString;
    class var kNXOAuth2AccountStoreConfigurationCustomHeaderFields: NSString;
    class var kNXOAuth2AccountStoreAccountType: NSString;
    class var NXOAuth2AccountDidChangeUserDataNotification: NSString;
    class var NXOAuth2AccountDidChangeAccessTokenNotification: NSString;
    class var NXOAuth2AccountDidLoseAccessTokenNotification: NSString;
    class var NXOAuth2AccountDidFailToGetAccessTokenNotification: NSString;
    class const NXOAuth2ConnectionDebug: Int32;
  end;

  OAuth2Client.NXOAuth2TrustMode = public enum (
    NXOAuth2TrustModeAnyCertificate,
    AnyCertificate,
    NXOAuth2TrustModeSystem,
    System,
    NXOAuth2TrustModeSpecificCertificate,
    SpecificCertificate
  );

  OAuth2Client.INXOAuth2TrustDelegate = public interface
    method connection(connection: OAuth2Client.NXOAuth2Connection) trustModeForHostname(hostname: NSString): OAuth2Client.NXOAuth2TrustMode;
    method connection(connection: OAuth2Client.NXOAuth2Connection) trustedCertificatesForHostname(hostname: NSString): NSArray;
  end;

  OAuth2Client.INXOAuth2ClientDelegate = public interface
    method oauthClientNeedsAuthentication(client: OAuth2Client.NXOAuth2Client);
    method oauthClientDidGetAccessToken(client: OAuth2Client.NXOAuth2Client);
    method oauthClientDidLoseAccessToken(client: OAuth2Client.NXOAuth2Client);
    method oauthClientDidRefreshAccessToken(client: OAuth2Client.NXOAuth2Client);
    method oauthClient(client: OAuth2Client.NXOAuth2Client) didFailToGetAccessTokenWithError(error: NSError);
  end;

  OAuth2Client.INXOAuth2ConnectionDelegate = public interface
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didReceiveResponse(response: NSURLResponse);
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didFinishWithData(data: NSData);
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didFailWithError(error: NSError);
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didReceiveData(data: NSData);
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didSendBytes(bytesSend: UInt64) ofTotal(bytesTotal: UInt64);
    method oauthConnection(connection: OAuth2Client.NXOAuth2Connection) didReceiveRedirectToURL(redirectURL: NSURL);
  end;

  OAuth2Client.NXOAuth2Client = public class(NSObject)
  protected
    var authenticating: BOOL;
    var persistent: BOOL;
    var clientId: NSString;
    var clientSecret: NSString;
    var desiredScope: NSSet;
    var userAgent: NSString;
    var assertion: NSString;
    var keyChainGroup: NSString;
    var authorizeURL: NSURL;
    var tokenURL: NSURL;
    var tokenType: NSString;
    var authConnection: OAuth2Client.NXOAuth2Connection;
    var accessToken: OAuth2Client.NXOAuth2AccessToken;
    var waitingConnections: NSMutableArray;
    var refreshConnectionDidRetryCount: NSInteger;
    var &delegate: OAuth2Client.INXOAuth2ClientDelegate;
  public
    property authenticating: BOOL read;
    property clientId: NSString read;
    property clientSecret: NSString read;
    property tokenType: NSString read;
    property additionalAuthenticationParameters: NSDictionary read write;
    property customHeaderFields: NSDictionary read write;
    property desiredScope: NSSet read write;
    property tokenRequestHTTPMethod: NSString read write;
    property userAgent: NSString read write;
    property acceptType: NSString read write;
    property accessToken: OAuth2Client.NXOAuth2AccessToken read write;
    property &delegate: OAuth2Client.INXOAuth2ClientDelegate read write;
    property persistent: BOOL read write;
    method initWithClientID(clientId: NSString) clientSecret(clientSecret: NSString) authorizeURL(authorizeURL: NSURL) tokenURL(tokenURL: NSURL) &delegate(&delegate: OAuth2Client.INXOAuth2ClientDelegate): id;
    method initWithClientID(clientId: NSString) clientSecret(clientSecret: NSString) authorizeURL(authorizeURL: NSURL) tokenURL(tokenURL: NSURL) accessToken(accessToken: OAuth2Client.NXOAuth2AccessToken) keyChainGroup(keyChainGroup: NSString) persistent(shouldPersist: BOOL) &delegate(&delegate: OAuth2Client.INXOAuth2ClientDelegate): id;
    method initWithClientID(clientId: NSString) clientSecret(clientSecret: NSString) authorizeURL(authorizeURL: NSURL) tokenURL(tokenURL: NSURL) accessToken(accessToken: OAuth2Client.NXOAuth2AccessToken) tokenType(tokenType: NSString) keyChainGroup(keyChainGroup: NSString) persistent(shouldPersist: BOOL) &delegate(&delegate: OAuth2Client.INXOAuth2ClientDelegate): id;
    method openRedirectURL(URL: NSURL): BOOL;
    method authorizationURLWithRedirectURL(redirectURL: NSURL): NSURL;
    method authenticateWithClientCredentials;
    method authenticateWithUsername(username: NSString) password(password: NSString);
    method authenticateWithAssertionType(assertionType: NSURL) assertion(assertion: NSString);
    method requestAccess;
    method refreshAccessToken;
    method refreshAccessTokenAndRetryConnection(retryConnection: OAuth2Client.NXOAuth2Connection);
  end;

  OAuth2Client.NXOAuth2AccessToken = public class(NSObject)
  private
    var accessToken: NSString;
    var refreshToken: NSString;
    var tokenType: NSString;
    var expiresAt: NSDate;
    var scope: NSSet;
    var responseBody: NSString;
  public
    property accessToken: NSString read;
    property refreshToken: NSString read;
    property tokenType: NSString read;
    property expiresAt: NSDate read;
    property doesExpire: BOOL read;
    property hasExpired: BOOL read;
    property scope: NSSet read;
    property responseBody: NSString read;
    class method tokenWithResponseBody(responseBody: NSString): id;
    class method tokenWithResponseBody(responseBody: NSString) tokenType(tokenType: NSString): id;
    method initWithAccessToken(accessToken: NSString): id;
    method initWithAccessToken(accessToken: NSString) refreshToken(refreshToken: NSString) expiresAt(expiryDate: NSDate): id;
    method initWithAccessToken(accessToken: NSString) refreshToken(refreshToken: NSString) expiresAt(expiryDate: NSDate) scope(scope: NSSet): id;
    method initWithAccessToken(accessToken: NSString) refreshToken(refreshToken: NSString) expiresAt(expiryDate: NSDate) scope(scope: NSSet) responseBody(responseBody: NSString): id;
    method initWithAccessToken(accessToken: NSString) refreshToken(refreshToken: NSString) expiresAt(expiryDate: NSDate) scope(scope: NSSet) responseBody(responseBody: NSString) tokenType(tokenType: NSString): id;
    method restoreWithOldToken(oldToken: OAuth2Client.NXOAuth2AccessToken);
    class method tokenFromDefaultKeychainWithServiceProviderName(provider: NSString): id;
    method storeInDefaultKeychainWithServiceProviderName(provider: NSString);
    method removeFromDefaultKeychainWithServiceProviderName(provider: NSString);
  end;

  OAuth2Client.NXOAuth2ConnectionResponseHandler = public method (response: NSURLResponse; responseData: NSData; error: NSError);

  OAuth2Client.NXOAuth2ConnectionSendingProgressHandler = public method (bytesSend: UInt64; bytesTotal: UInt64);

  OAuth2Client.NXOAuth2Connection = public class(NSObject)
  private
    var connection: NSURLConnection;
    var request: NSMutableURLRequest;
    var response: NSURLResponse;
    var requestParameters: NSDictionary;
    var data: NSMutableData;
    var savesData: BOOL;
    var context: id;
    var userInfo: NSDictionary;
    var client: OAuth2Client.NXOAuth2Client;
    var &delegate: OAuth2Client.INXOAuth2ConnectionDelegate;
    var responseHandler: OAuth2Client.NXOAuth2ConnectionResponseHandler;
    var sendingProgressHandler: OAuth2Client.NXOAuth2ConnectionSendingProgressHandler;
    var sendConnectionDidEndNotification: BOOL;
  public
    property &delegate: OAuth2Client.INXOAuth2ConnectionDelegate read write;
    property data: NSData read;
    property savesData: BOOL read write;
    property expectedContentLength: Int64 read;
    property response: NSURLResponse read;
    property statusCode: NSInteger read;
    property context: id read write;
    property userInfo: NSDictionary read write;
    property client: OAuth2Client.NXOAuth2Client read;
    method initWithRequest(request: NSMutableURLRequest) requestParameters(requestParameters: NSDictionary) oauthClient(client: OAuth2Client.NXOAuth2Client) sendingProgressHandler(sendingProgressHandler: OAuth2Client.NXOAuth2ConnectionSendingProgressHandler) responseHandler(responseHandler: OAuth2Client.NXOAuth2ConnectionResponseHandler): id;
    method initWithRequest(request: NSMutableURLRequest) requestParameters(requestParameters: NSDictionary) oauthClient(client: OAuth2Client.NXOAuth2Client) &delegate(&delegate: OAuth2Client.INXOAuth2ConnectionDelegate): id;
    method cancel;
    method retry;
  end;

  OAuth2Client.NXOAuth2FileStreamWrapper = public class(NSObject)
  protected
    var stream: NSInputStream;
    var contentLength: UInt64;
    var fileName: NSString;
    var contentType: NSString;
  public
    property stream: NSInputStream read;
    property contentLength: UInt64 read;
    property fileName: NSString read;
    property contentType: NSString read write;
    class method wrapperWithStream(stream: NSInputStream) contentLength(contentLength: UInt64): id;
    method initWithStream(stream: NSInputStream) contentLength(contentLength: UInt64): id;
    class method wrapperWithStream(stream: NSInputStream) contentLength(contentLength: UInt64) fileName(fileName: NSString): id;
    method initWithStream(stream: NSInputStream) contentLength(contentLength: UInt64) fileName(fileName: NSString): id;
  end;

  OAuth2Client.NXOAuth2PostBodyStream = public class(NSInputStream)
  protected
    var boundary: NSString;
    var contentStreams: NSArray;
    var currentStream: NSInputStream;
    var streamIndex: NSUInteger;
    var numBytesTotal: UInt64;
  public
    method initWithParameters(postParameters: NSDictionary): id;
    property boundary: NSString read;
    property length: UInt64 read;
  end;

  OAuth2Client.NXOAuth2TrustModeHandler = public method (connection: OAuth2Client.NXOAuth2Connection; hostname: NSString): OAuth2Client.NXOAuth2TrustMode;

  OAuth2Client.NXOAuth2TrustedCertificatesHandler = public method (hostname: NSString): NSArray;

  OAuth2Client.NXOAuth2PreparedAuthorizationURLHandler = public method (preparedURL: NSURL);

  OAuth2Client.NXOAuth2AccountStore = public class(NSObject)
  private
    var pendingOAuthClients: NSMutableDictionary;
    var accountsDict: NSMutableDictionary;
    var configurations: NSMutableDictionary;
    var trustModeHandler: NSMutableDictionary;
    var trustedCertificatesHandler: NSMutableDictionary;
  public
    class method sharedStore: id;
    property accounts: NSArray read;
    method accountsWithAccountType(accountType: NSString): NSArray;
    method accountWithIdentifier(identifier: NSString): OAuth2Client.NXOAuth2Account;
    method setClientID(aClientID: NSString) secret(aSecret: NSString) authorizationURL(anAuthorizationURL: NSURL) tokenURL(aTokenURL: NSURL) redirectURL(aRedirectURL: NSURL) forAccountType(anAccountType: NSString);
    method setClientID(aClientID: NSString) secret(aSecret: NSString) scope(theScope: NSSet) authorizationURL(anAuthorizationURL: NSURL) tokenURL(aTokenURL: NSURL) redirectURL(aRedirectURL: NSURL) keyChainGroup(aKeyChainGroup: NSString) forAccountType(anAccountType: NSString);
    method setClientID(aClientID: NSString) secret(aSecret: NSString) scope(theScope: NSSet) authorizationURL(anAuthorizationURL: NSURL) tokenURL(aTokenURL: NSURL) redirectURL(aRedirectURL: NSURL) keyChainGroup(aKeyChainGroup: NSString) tokenType(aTokenType: NSString) forAccountType(anAccountType: NSString);
    method setConfiguration(configuration: NSDictionary) forAccountType(accountType: NSString);
    method configurationForAccountType(accountType: NSString): NSDictionary;
    method setTrustModeHandlerForAccountType(accountType: NSString) &block(handler: OAuth2Client.NXOAuth2TrustModeHandler);
    method trustModeHandlerForAccountType(accountType: NSString): OAuth2Client.NXOAuth2TrustModeHandler;
    method setTrustedCertificatesHandlerForAccountType(accountType: NSString) &block(handler: OAuth2Client.NXOAuth2TrustedCertificatesHandler);
    method trustedCertificatesHandlerForAccountType(accountType: NSString): OAuth2Client.NXOAuth2TrustedCertificatesHandler;
    method requestAccessToAccountWithType(accountType: NSString);
    method requestAccessToAccountWithType(accountType: NSString) withPreparedAuthorizationURLHandler(aPreparedAuthorizationURLHandler: OAuth2Client.NXOAuth2PreparedAuthorizationURLHandler);
    method requestAccessToAccountWithType(accountType: NSString) username(username: NSString) password(password: NSString);
    method requestAccessToAccountWithType(accountType: NSString) assertionType(assertionType: NSURL) assertion(assertion: NSString);
    method requestClientCredentialsAccessWithType(accountType: NSString);
    method addAccount(account: OAuth2Client.NXOAuth2Account);
    method removeAccount(account: OAuth2Client.NXOAuth2Account);
    method handleRedirectURL(URL: NSURL): BOOL;
  end;

  OAuth2Client.NXOAuth2Account = public class(NSObject)
  private
    var accountType: NSString;
    var identifier: NSString;
    var userData: <anonymous type>;
    var oauthClient: OAuth2Client.NXOAuth2Client;
    var accessToken: OAuth2Client.NXOAuth2AccessToken;
  public
    property accountType: NSString read;
    property identifier: NSString read;
    property userData: <anonymous type> read write;
    property oauthClient: OAuth2Client.NXOAuth2Client read;
    property accessToken: OAuth2Client.NXOAuth2AccessToken read;
  end;

  OAuth2Client.NXOAuth2Request = public class(NSObject)
  private
    var parameters: NSDictionary;
    var resource: NSURL;
    var requestMethod: NSString;
    var account: OAuth2Client.NXOAuth2Account;
    var connection: OAuth2Client.NXOAuth2Connection;
    var me: OAuth2Client.NXOAuth2Request;
  public
    class method performMethod(&method: NSString) onResource(resource: NSURL) usingParameters(parameters: NSDictionary) withAccount(account: OAuth2Client.NXOAuth2Account) sendProgressHandler(progressHandler: OAuth2Client.NXOAuth2ConnectionSendingProgressHandler) responseHandler(responseHandler: OAuth2Client.NXOAuth2ConnectionResponseHandler);
    method initWithResource(url: NSURL) &method(&method: NSString) parameters(parameter: NSDictionary): id;
    property account: OAuth2Client.NXOAuth2Account read write;
    property requestMethod: NSString read write;
    property resource: NSURL read write;
    property parameters: NSDictionary read write;
    method signedURLRequest: NSURLRequest;
    method performRequestWithSendingProgressHandler(progressHandler: OAuth2Client.NXOAuth2ConnectionSendingProgressHandler) responseHandler(responseHandler: OAuth2Client.NXOAuth2ConnectionResponseHandler);
    method cancel;
  end;

implementation

end.
